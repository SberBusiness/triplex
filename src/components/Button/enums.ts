/** Темы кнопки. */
export enum EButtonTheme {
    GENERAL = 'general',
    SECONDARY = 'secondary',
    DANGER = 'danger',
    DOTS = 'dots',
    LINK = 'link',
    TILE = 'tile',
}

/** Размеры кнопки. */
export enum EButtonSize {
    LG = 'LG',
    MD = 'MD',
    SM = 'SM',
}

/** Форма границы кнопки. */
export enum EButtonIconShape {
    SQUIRCLE = 'squircle',
    CIRCLE = 'circle',
}
