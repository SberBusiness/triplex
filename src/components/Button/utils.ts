import React from 'react';

/**
 * Не происходит фокус при клике на button в сафари и в мозиле на маках.
 * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
 *
 * Данная функция должна вызываться только синхронно. (Не будет работать в колбэках setState)
 * https://stackoverflow.com/questions/42089795/reactjs-cant-set-state-from-an-event-with-event-persist
 */
export const focusButton = (event: React.MouseEvent<HTMLButtonElement>): void => {
    event.currentTarget?.focus();
};
