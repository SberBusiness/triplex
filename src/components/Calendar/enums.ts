/** Вариант выбора даты. */
export enum EPickType {
    datePick,
    monthYearPick,
}

/**
 * Направление переключения страницы календаря.
 */
export enum EChangePageDirection {
    FORWARD,
    BACKWARD,
}
