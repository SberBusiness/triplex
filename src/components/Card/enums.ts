/** Возможные размеры внутреннего отступа. */
export enum ECardContentPaddingSize {
    MD = 'md',
    SM = 'sm',
}

/** Возможные размеры скругления карточки. */
export enum ECardRoundingSize {
    MD = 'md',
    SM = 'sm',
}
