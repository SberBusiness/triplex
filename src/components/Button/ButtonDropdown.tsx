import React, {useRef, useState} from 'react';
import {
    ButtonDropdownExtended,
    IButtonDropdownExtendedButtonProvideProps,
    IButtonDropdownExtendedDropdownProvideProps,
} from '@sberbusiness/triplex/components/Button/ButtonDropdownExtended';
import {Button} from '@sberbusiness/triplex/components/Button/Button';
import {EButtonSize, EButtonTheme} from '@sberbusiness/triplex/components/Button/enums';
import {DropdownwhiteSrvxIcon16} from '@sberbusiness/icons/DropdownwhiteSrvxIcon16';
import {CaretdownSrvxIcon16} from '@sberbusiness/icons/CaretdownSrvxIcon16';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {isKey} from '@sberbusiness/triplex/utils/keyboard';
import {DropdownList} from '@sberbusiness/triplex/components/Dropdown/desktop/DropdownList';
import {IDropdownListItemProps} from '../Dropdown/desktop/DropdownListItem';
import {DropdownListContext} from '@sberbusiness/triplex/components/Dropdown/DropdownListContext';
import {uniqueId} from '@sberbusiness/triplex/utils/uniqueId';
import {DropdownMobileHeader} from '@sberbusiness/triplex/components/Dropdown/mobile/DropdownMobileHeader';
import {DropdownMobileBody} from '@sberbusiness/triplex/components/Dropdown/mobile/DropdownMobileBody';
import {DropdownMobileList} from '@sberbusiness/triplex/components/Dropdown/mobile/DropdownMobileList';
import {DropdownMobileListItem} from '@sberbusiness/triplex/components/Dropdown/mobile/DropdownMobileListItem';
import {DropdownMobileClose} from '@sberbusiness/triplex/components/Dropdown/mobile/DropdownMobileClose';
import {Text} from '@sberbusiness/triplex/components/Typography/Text';
import {ELineType, ETextSize} from '@sberbusiness/triplex/components/Typography/enums';

/** Свойства опции в выпадающем списке действий. */
export interface IButtonDropdownOption
    extends Omit<IDropdownListItemProps, 'active' | 'onSelect' | 'selected' | 'keyCodesForSelection' | 'className' | 'key'> {
    /** Уникальное имя опции. */
    id: string;
    /** Название опции. */
    label: React.ReactNode;
    /** Обработчик выбора опции. */
    onSelect?: () => void;
}

/** Свойства кнопки с выпадающим списком действий. */
interface IButtonDropdownProps extends React.HTMLAttributes<HTMLDivElement> {
    /** HTML-атрибуты кнопки. */
    buttonAttributes?: React.ButtonHTMLAttributes<HTMLButtonElement>;
    /** Размер кнопки. */
    size: EButtonSize;
    /** Список опций. */
    options: IButtonDropdownOption[];
    /** Выбранная опция. */
    selected?: IButtonDropdownOption;
    /** Отключенное состояние кнопки. */
    disabled?: boolean;
}

/** Свойства основной/вспомогательной кнопки с выпадающим списком действий. */
interface IButtonDropdownBaseProps extends IButtonDropdownProps {
    /** Тема кнопки. */
    theme: EButtonTheme.GENERAL | EButtonTheme.SECONDARY | EButtonTheme.DANGER;
    /** Блочное состояние кнопки. */
    block?: boolean;
}

/** Свойства контекстной кнопки с выпадающим списком действий. */
interface IButtonDotsProps extends IButtonDropdownProps {
    /** Тема кнопки. */
    theme: EButtonTheme.DOTS;
    /** Блочное состояние кнопки. */
    block?: never;
}

/** Кнопка с выпадающим списком действий. */
export const ButtonDropdown = React.forwardRef<HTMLButtonElement, IButtonDropdownBaseProps | IButtonDotsProps>((props, ref) => {
    const {buttonAttributes, children, className, theme, size, options, selected, block, disabled, ...rest} = props;
    const buttonRef = useRef<HTMLButtonElement | null>(null);
    const dropdownRef = useRef<HTMLDivElement>(null);
    const classNames = classnames('cssClass[buttonDropdown]', {'cssClass[block]': !!block}, className);
    const [activeDescendant, setActiveDescendant] = useState<string>();
    const instanceId = useRef(uniqueId());

    const renderButton = ({opened, setOpened}: IButtonDropdownExtendedButtonProvideProps) => {
        const classNames = classnames('cssClass[buttonDropdownTarget]', 'hoverable', {
            'cssClass[active]': opened,
            'cssClass[block]': !!block,
        });

        return (
            <Button
                className={classNames}
                theme={theme}
                size={size}
                onKeyDown={handleKeyDown({opened, setOpened})}
                onClick={handleClick({opened, setOpened})}
                disabled={disabled}
                aria-haspopup="menu"
                aria-expanded={opened}
                aria-controls={instanceId.current}
                aria-activedescendant={activeDescendant}
                {...buttonAttributes}
                ref={setRef}
            >
                {children}
                {renderCaret()}
            </Button>
        );
    };

    const handleClick =
        ({opened, setOpened}: IButtonDropdownExtendedButtonProvideProps) =>
        () =>
            setOpened(!opened);

    const handleKeyDown =
        ({opened, setOpened}: IButtonDropdownExtendedButtonProvideProps) =>
        (event: React.KeyboardEvent<HTMLButtonElement>) => {
            const {key} = event;

            if (isKey(key, 'SPACE') || isKey(key, 'ARROW_UP') || isKey(key, 'ARROW_DOWN')) {
                event.preventDefault();
            }
            if (!opened && (isKey(key, 'ARROW_UP') || isKey(key, 'ARROW_DOWN'))) {
                setOpened(true);
            }
        };

    const renderCaret = () => {
        switch (theme) {
            case EButtonTheme.GENERAL:
            case EButtonTheme.DANGER:
                return <DropdownwhiteSrvxIcon16 className="cssClass[caretIcon]" />;
            case EButtonTheme.SECONDARY:
                return <CaretdownSrvxIcon16 className="cssClass[caretIcon]" />;
        }
    };

    const renderDropdown = ({opened, setOpened, className}: IButtonDropdownExtendedDropdownProvideProps) => {
        const classNames = classnames('cssClass[buttonDropdownMenu]', className);

        return (
            <DropdownListContext.Provider value={{activeDescendant, setActiveDescendant}}>
                <ButtonDropdownExtended.Dropdown
                    className={classNames}
                    opened={opened}
                    setOpened={setOpened}
                    targetRef={buttonRef}
                    ref={dropdownRef}
                    mobileViewProps={{
                        children: (
                            <>
                                <DropdownMobileHeader>
                                    <Text size={ETextSize.B1} line={ELineType.EXTRA}>
                                        {children}
                                    </Text>
                                    <DropdownMobileClose onClick={() => setOpened(false)} />
                                </DropdownMobileHeader>
                                <DropdownMobileBody>
                                    <DropdownMobileList>
                                        {options.map((option) => (
                                            <DropdownMobileListItem
                                                {...option}
                                                key={option.id}
                                                selected={option.id === selected?.id}
                                                onSelect={() => {
                                                    option.onSelect?.();
                                                    setOpened(false);
                                                }}
                                            >
                                                {option.label}
                                            </DropdownMobileListItem>
                                        ))}
                                    </DropdownMobileList>
                                </DropdownMobileBody>
                            </>
                        ),
                    }}
                >
                    <DropdownList dropdownOpened={opened} id={instanceId.current}>
                        {options.map((option) => (
                            <DropdownList.Item
                                {...option}
                                className="cssClass[buttonDropdownMenuItem]"
                                key={option.id}
                                selected={option.id === selected?.id}
                                onSelect={() => {
                                    option.onSelect?.();
                                    setOpened(false);
                                }}
                            >
                                {option.label}
                            </DropdownList.Item>
                        ))}
                    </DropdownList>
                </ButtonDropdownExtended.Dropdown>
            </DropdownListContext.Provider>
        );
    };

    /** Функция для хранения ссылки. */
    const setRef = (instance: HTMLButtonElement | null) => {
        buttonRef.current = instance;
        if (typeof ref === 'function') {
            ref(instance);
        } else if (ref) {
            ref.current = instance;
        }
    };

    return (
        <ButtonDropdownExtended
            className={classNames}
            renderButton={renderButton}
            renderDropdown={renderDropdown}
            dropdownRef={dropdownRef}
            closeOnTab
            {...rest}
        />
    );
});

ButtonDropdown.displayName = 'ButtonDropdown';
