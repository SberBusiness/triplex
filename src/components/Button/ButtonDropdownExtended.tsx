import React from 'react';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {Dropdown} from '@sberbusiness/triplex/components/Dropdown/Dropdown';
import {isKey} from '@sberbusiness/triplex/utils/keyboard';
import {DropdownList} from '@sberbusiness/triplex/components/Dropdown/desktop/DropdownList';

/** Состояния кнопки с выпадающим блоком. */
interface IButtonDropdownExtendedState {
    /** Состояние контролируемости. */
    isControlled: boolean;
    /** Состояние открытости. */
    isOpened?: boolean;
}

/** Свойства встроенной кнопки. */
export interface IButtonDropdownExtendedButtonProvideProps {
    /** Контролируемое состояние открытости. */
    opened: boolean;
    /** Функция, контролирующая состояние открытости. */
    setOpened: (opened: boolean) => void;
}

/** Свойства встроенного выпадающего блока. */
export interface IButtonDropdownExtendedDropdownProvideProps {
    /** Контролируемое состояние открытости. */
    opened: boolean;
    /** Функция, контролирующая состояние открытости. */
    setOpened: (opened: boolean) => void;
    /** Пробрасываемый стилевой класс. */
    className: string;
}

/** Свойства кнопки с выпадающим блоком. */
export interface IButtonDropdownExtendedProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Контролируемое состояние открытости. */
    opened?: boolean;
    /** Функция, контролирующая состояние открытости. */
    setOpened?: (opened: boolean) => void;
    /** Функция, отрисовывающая кнопку. */
    renderButton: (props: IButtonDropdownExtendedButtonProvideProps) => React.ReactNode;
    /** Функция, отрисовывающая выпадающий блок. */
    renderDropdown: (props: IButtonDropdownExtendedDropdownProvideProps) => React.ReactNode;
    /** Ссылка на выпадающий блок. */
    dropdownRef: React.RefObject<HTMLElement>;
    /** Закрытие выпадающего блока при нажатии клавиши Tab. */
    closeOnTab?: boolean;
}

/**
 * Компонент "Кнопка с выпадающим блоком".
 * Позволяет кастомизировать кнопку открытия Dropdown и сам Dropdown.
 * */
export class ButtonDropdownExtended extends React.Component<IButtonDropdownExtendedProps, IButtonDropdownExtendedState> {
    public static Dropdown = Dropdown;
    public static DropdownList = DropdownList;

    constructor(props: IButtonDropdownExtendedProps) {
        super(props);

        this.state = {isControlled: props.opened !== undefined, isOpened: props.opened === undefined ? false : undefined};
    }

    private ref = React.createRef<HTMLDivElement>();

    private getOpened = (): boolean => {
        const {opened} = this.props;
        const {isControlled, isOpened} = this.state;

        return !!(isControlled ? opened : isOpened);
    };

    private handleOpen = (opened: boolean) => {
        const {setOpened} = this.props;
        const {isControlled} = this.state;

        isControlled ? setOpened!(opened) : this.setState({isOpened: opened});
    };

    private handleKeyDown = (event: KeyboardEvent) => {
        const {closeOnTab} = this.props;
        const key = event.code || event.keyCode;

        if (this.getOpened()) {
            if (isKey(key, 'ESCAPE') || (closeOnTab && isKey(key, 'TAB'))) {
                this.handleOpen(false);
            }
        }
    };

    private handleClickOutside = (event: UIEvent) => {
        const {current: button} = this.ref;
        const {current: dropdown} = this.props.dropdownRef;

        if (this.getOpened()) {
            if (!button?.contains(event.target as Node) && !dropdown?.contains(event.target as Node)) {
                this.handleOpen(false);
            }
        }
    };

    private addListeners(): void {
        document.addEventListener('keydown', this.handleKeyDown);
        document.addEventListener('mousedown', this.handleClickOutside);
        document.addEventListener('touchstart', this.handleClickOutside);
    }

    private removeListeners(): void {
        document.removeEventListener('keydown', this.handleKeyDown);
        document.removeEventListener('mousedown', this.handleClickOutside);
        document.removeEventListener('touchstart', this.handleClickOutside);
    }

    public componentDidMount(): void {
        if (this.getOpened()) {
            this.addListeners();
        }
    }

    public componentDidUpdate(prevProps: Readonly<IButtonDropdownExtendedProps>, prevState: Readonly<IButtonDropdownExtendedState>): void {
        const {opened} = this.props;
        const {isOpened, isControlled} = this.state;
        const {opened: prevOpened} = prevProps;
        const {isOpened: prevIsOpened} = prevState;

        if (isControlled) {
            if (opened && !prevOpened) {
                this.addListeners();
            } else if (!opened && prevOpened) {
                this.removeListeners();
            }
        } else {
            if (isOpened && !prevIsOpened) {
                this.addListeners();
            } else if (!isOpened && prevIsOpened) {
                this.removeListeners();
            }
        }
    }

    public componentWillUnmount(): void {
        if (this.getOpened()) {
            this.removeListeners();
        }
    }

    public render(): JSX.Element {
        const {className, opened, setOpened, renderButton, renderDropdown, dropdownRef, closeOnTab, ...props} = this.props;
        const classNames = classnames('cssClass[globalButtonDropdownExtended]', className);

        return (
            <div className={classNames} ref={this.ref} {...props}>
                {renderButton({opened: this.getOpened(), setOpened: this.handleOpen})}
                {renderDropdown({className: 'cssClass[buttonDropdownExtendedBlock]', opened: this.getOpened(), setOpened: this.handleOpen})}
            </div>
        );
    }
}
