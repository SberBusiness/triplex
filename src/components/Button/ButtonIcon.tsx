import React from 'react';
import {EButtonIconShape} from '@sberbusiness/triplex/components/Button/enums';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {ButtonBase} from '../protected/ButtonBase/ButtonBase';

const getButtonIconShapeClassName = (shape?: EButtonIconShape) => {
    switch (shape) {
        case EButtonIconShape.SQUIRCLE:
            return 'cssClass[squircle]';
        case EButtonIconShape.CIRCLE:
            return 'cssClass[circle]';
    }
};

/** Свойства ButtonIcon. */
export interface IButtonIconProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    /** Форма границы кнопки. */
    shape?: EButtonIconShape;
    /** Активное состояние. */
    active?: boolean;
}

/** Кнопка-иконка. */
export const ButtonIcon = React.forwardRef<HTMLButtonElement, IButtonIconProps>(
    ({className, disabled, shape = EButtonIconShape.SQUIRCLE, active, ...rest}, ref) => {
        const classNames = classnames('cssClass[buttonIcon]', getButtonIconShapeClassName(shape), 'hoverable', className, {
            active: !!active,
            disabled: !!disabled,
        });

        return <ButtonBase className={classNames} disabled={disabled} {...rest} ref={ref} />;
    }
);

ButtonIcon.displayName = 'ButtonIcon';

export {EButtonIconShape};
