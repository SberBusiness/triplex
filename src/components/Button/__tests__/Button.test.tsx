import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {Button} from '@sberbusiness/triplex/components/Button/Button';
import {EButtonTheme, EButtonSize} from '@sberbusiness/triplex/components/Button/enums';

jest.mock('@sberbusiness/icons/SpinnersmallAniIcon20', () => ({
    SpinnersmallAniIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/SpinnersmallwhiteAniIcon20', () => ({
    SpinnersmallwhiteAniIcon20: 'svg',
}));

describe('Button', () => {
    beforeEach(() => {
        allure.feature('Button');
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD}>
                    Button Name
                </Button>
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('triggers click event', () => {
        const handleClick = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onClick={handleClick} />);

        button.simulate('click');
        expect(handleClick).toHaveBeenCalled();
    });

    it('triggers mouseOver event', () => {
        const handleMouseOver = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onMouseOver={handleMouseOver} />);

        button.simulate('mouseOver');
        expect(handleMouseOver).toHaveBeenCalled();
    });

    it('triggers mouseOut event', () => {
        const handleMouseOut = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onMouseOut={handleMouseOut} />);

        button.simulate('mouseOut');
        expect(handleMouseOut).toHaveBeenCalled();
    });

    it('triggers mouseDown event', () => {
        const handleMouseDown = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onMouseDown={handleMouseDown} />);

        button.simulate('mouseDown');
        expect(handleMouseDown).toHaveBeenCalled();
    });

    it('triggers mouseUp event', () => {
        const handleMouseUp = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onMouseUp={handleMouseUp} />);

        button.simulate('mouseUp');
        expect(handleMouseUp).toHaveBeenCalled();
    });

    it('triggers keyDown event', () => {
        const handleKeyDown = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onKeyDown={handleKeyDown} />);

        button.simulate('keyDown');
        expect(handleKeyDown).toHaveBeenCalled();
    });

    it('triggers keyUp event', () => {
        const handleKeyUp = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onKeyUp={handleKeyUp} />);

        button.simulate('keyUp');
        expect(handleKeyUp).toHaveBeenCalled();
    });

    it('triggers focus event', () => {
        const handleFocus = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onFocus={handleFocus} />);

        button.simulate('focus');
        expect(handleFocus).toHaveBeenCalled();
    });

    it('triggers blur event', () => {
        const handleBlur = jest.fn();
        const button = shallow(<Button theme={EButtonTheme.GENERAL} size={EButtonSize.MD} onBlur={handleBlur} />);

        button.simulate('blur');
        expect(handleBlur).toHaveBeenCalled();
    });
});
