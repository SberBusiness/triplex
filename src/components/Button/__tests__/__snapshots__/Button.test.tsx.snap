// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  className="cssClass[button] cssClass[general] cssClass[MD]"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  type="button"
>
  <span
    className="cssClass[content]"
  >
    Button Name
  </span>
</button>
`;
